### ISSUE TITLE
Input with different lengths in DSSM_LSTM

### ISSUE THREAD

**Issue by secretsh** [microsoft/CNTK#3707]
Hi, 

I am trying to build a brainscript for DSSM_LSTM model similar to this one https://cntk.ai/pythondocs/CNTK_303_Deep_Structured_Semantic_Modeling_with_LSTM_Networks.html 

There could be many cases that the query and document are with different lengths, as following example:

0	|Query0 1160:1	|Query1 1234:1	|Doc0 956:1	|Doc1 234:1
0	|Query0 1111:1	|Query1 5678:1
1	|Query0 1160:1	|Query1 1234:1	|Doc0 956:1	|Doc1 234:1
1									|Doc0 956:1	|Doc1 234:1

In this case, what would happen to the side with shorter length? Is the input truncated? Or is the token with empty input will be ignored during in LSTM computation? 

Thank you

**Comment by secretsh** [comment]
Hi, 
I change my strategy to use two separate inputs of query and documents. Hopefully this will work.

**Comment by secretsh** [comment]
Hi, I got the error message like "Dynamic axis layout '*' is shared between inputs" using inputs like above;

Then I tried to add DynamicAxis() as following:

inputAxisQ = DynamicAxis()
inputAxisK = DynamicAxis()
		
query      = Input {inputDim, dynamicAxis=inputAxisQ}
doc= Input {inputDim, dynamicAxis=inputAxisK }

But then I got following error
"About to throw exception 'TensorSliceWithMBLayoutFor: FrameRange's dynamic axis is inconsistent with data: {numTimeSteps:3, numParallelSequences:2, sequences:[{seqId:0, s:0, begin:0, end:3}, {seqId:1, s:1, begin:0, end:3}]} vs. {numTimeSteps:2, numParallelSequences:2, sequences:[{seqId:0, s:0, begin:0, end:2}, {seqId:1, s:1, begin:0, end:2}]}'"

How can I make it right?

Thank you

**Comment by secretsh** [comment]
Hi, I am able to read the data with using DynamicAxis

My next problem is that, how to get the average pooling of all tokens after LSTM computation?

The operator AveragePoolingLayer seems like only able to get the average values of vector from one single token. 

Thank you

**Comment by delzac** [comment]
For your first question, it means like your input 0 doesn't have a dynamic batch axis.


For your second question on how to implement a sequence.reduce_mean, i made one in the cntk extension library that i maintain called [cntkx](https://github.com/delzac/cntkx). You can use it as a reference to implement the same in BrainScript.

**Comment by secretsh** [comment]
Hi thank you for your reply.

For the first question, I tried to add DynamicAxis like this

inputAxisK = DynamicAxis()
inputAxisQ = DynamicAxis()
		
query      = Input {inputDim,dynamicAxis=inputAxisQ}
embed_     = Input {$fixemb_dim$,dynamicAxis=inputAxisQ}
slotLabels = Input {labelDim,dynamicAxis=inputAxisQ}
		
keyworrd = Input {inputDim,dynamicAxis=inputAxisK}


The error message remains the same. Am I adding the right thing?

Thank you!

**Comment by delzac** [comment]
Based on the exception raised, it seems like the data that you feed into the tensor doesn't have a minibatch. I probably isn't a DynamicAxis instantiation problem.

**Comment by secretsh** [comment]
Thank you @delzac , I think you are right.

After some debugging, I found it's because my training target "lf" does not have the same shape as the prediction "c", where "lf" is a single constant vector, and "c" is prediction**s** of a minibatch.

so I change the target to this:
labels = BS.Sequences.Last (slotLabels)
ce  = CrossEntropyWithSoftmax(labels ,dc , tag="criterion")

(I also set the label_size equals to (1 +negative sample ), so that the labels will be like (1, 0, 0, ...) which can be used for SOFTMAX between positive sample and negative samples)

Now the label and prediction will align in the shape, but I guess there could be better way of doing this.

My training can run now, I just don't know if it could run as expected...

**Comment by delzac** [comment]
Looks fine to me
